Mathlib/Data/PNat/Xgcd.lean

```diff
@@ -0,0 +1,556 @@
+  y : ℕ
+  /-- `zp` is a variable which changes through the algorithm. -/
+  zp : ℕ
+  /-- `ap` is a variable which changes through the algorithm. -/
+  ap : ℕ
+  /-- `bp` is a variable which changes through the algorithm. -/
+  bp : ℕ
+  deriving Inhabited
+#align pnat.xgcd_type PNat.XgcdType
+
+namespace XgcdType
+
+variable (u : XgcdType)
+
+instance : SizeOf XgcdType :=
+  ⟨fun u => u.bp⟩
+
+/-- The `Repr` instance converts terms to strings in a way that
+ reflects the matrix/vector interpretation as above. -/
+instance : Repr XgcdType where
+  reprPrec
+  | g, _ => s!"[[[ {repr (g.wp + 1)}, {(repr g.x)} ], [" ++
+            s!"{repr g.y}, {repr (g.zp + 1)}]], [" ++
+            s!"{repr (g.ap + 1)}, {repr (g.bp + 1)}]]"
+
+/-- Another `mk` using ℕ and ℕ+ -/
+def mk' (w : ℕ+) (x : ℕ) (y : ℕ) (z : ℕ+) (a : ℕ+) (b : ℕ+) : XgcdType :=
+  mk w.val.pred x y z.val.pred a.val.pred b.val.pred
+#align pnat.xgcd_type.mk' PNat.XgcdType.mk'
+
+/-- `w = wp + 1` -/
+def w : ℕ+ :=
+  succPNat u.wp
+#align pnat.xgcd_type.w PNat.XgcdType.w
+
+/-- `z = zp + 1` -/
+def z : ℕ+ :=
+  succPNat u.zp
+#align pnat.xgcd_type.z PNat.XgcdType.z
+
+/-- `a = ap + 1` -/
+def a : ℕ+ :=
+  succPNat u.ap
+#align pnat.xgcd_type.a PNat.XgcdType.a
+
+/-- `b = bp + 1` -/
+def b : ℕ+ :=
+  succPNat u.bp
+#align pnat.xgcd_type.b PNat.XgcdType.b
+
+/-- `r = a % b`: remainder -/
+def r : ℕ :=
+  (u.ap + 1) % (u.bp + 1)
+#align pnat.xgcd_type.r PNat.XgcdType.r
+
+/-- `q = ap / bp`: quotient -/
+def q : ℕ :=
+  (u.ap + 1) / (u.bp + 1)
+#align pnat.xgcd_type.q PNat.XgcdType.q
+
+/-- `qp = q - 1` -/
+def qp : ℕ :=
+  u.q - 1
+#align pnat.xgcd_type.qp PNat.XgcdType.qp
+
+/-- The map `v` gives the product of the matrix
+ [[w, x], [y, z]] = [[wp + 1, x], [y, zp + 1]]
+ and the vector [a, b] = [ap + 1, bp + 1].  The map
+ `vp` gives [sp, tp] such that v = [sp + 1, tp + 1].
+-/
+def vp : ℕ × ℕ :=
+  ⟨u.wp + u.x + u.ap + u.wp * u.ap + u.x * u.bp, u.y + u.zp + u.bp + u.y * u.ap + u.zp * u.bp⟩
+#align pnat.xgcd_type.vp PNat.XgcdType.vp
+
+/-- `v = [sp + 1, tp + 1]`, check `vp` -/
+def v : ℕ × ℕ :=
+  ⟨u.w * u.a + u.x * u.b, u.y * u.a + u.z * u.b⟩
+#align pnat.xgcd_type.v PNat.XgcdType.v
+
+/-- `succ₂ [t.1, t.2] = [t.1.succ, t.2.succ]` -/
+def succ₂ (t : ℕ × ℕ) : ℕ × ℕ :=
+  ⟨t.1.succ, t.2.succ⟩
+#align pnat.xgcd_type.succ₂ PNat.XgcdType.succ₂
+
+theorem v_eq_succ_vp : u.v = succ₂ u.vp := by
+  ext <;> dsimp [v, vp, w, z, a, b, succ₂] <;> (repeat' rw [Nat.succ_eq_add_one]; ring_nf)
+#align pnat.xgcd_type.v_eq_succ_vp PNat.XgcdType.v_eq_succ_vp
+
+/-- `isSpecial` holds if the matrix has determinant one. -/
+def isSpecial : Prop :=
+  u.wp + u.zp + u.wp * u.zp = u.x * u.y
+#align pnat.xgcd_type.is_special PNat.XgcdType.isSpecial
+
+/-- `isSpecial'` is an alternative of `isSpecial`. -/
+def isSpecial' : Prop :=
+  u.w * u.z = succPNat (u.x * u.y)
+#align pnat.xgcd_type.is_special' PNat.XgcdType.isSpecial'
+
+theorem is_special_iff : u.isSpecial ↔ u.isSpecial' := by
```

Comment:
Please fix this name. If you use VSCode, do the following:

1. select `is_special`
2. run the following loop:
  - hit `Ctrl-D` to select the next occurence
  - if the occurence should be skipped, hit `Ctrl-K` to "kill" that selection
  - (if necessary, hit `Ctrl-U` to backtrack)
3. after you have selected all occurences of `is_special` that need to be changed, make the edit to `isSpecial`.